File: src/purra/approvals.py
"""Immutable durable approval data. These records never grant dispatch authority."""
from dataclasses import dataclass, field
from typing import Any, Mapping

from purra.json_values import freeze_json_mapping, thaw_json_mapping
from purra.structured import StructuredOutputContract, json_identity_digest


APPROVAL_INTENT_PROFILE = "purra.approval-intent/v1"
APPROVAL_STATUSES = frozenset({"pending", "approved", "rejected", "expired", "canceled"})
_ARGUMENTS = StructuredOutputContract("purra.approval-arguments", "1", {"type": "object"})


def _text(value, name):
    if not isinstance(value, str) or not value or value.strip() != value or len(value) > 1024:
        raise ValueError(f"invalid approval {name}")
    return value


def _integer(value, name, minimum=0):
    if type(value) is not int or not minimum <= value <= 9007199254740991:
        raise ValueError(f"invalid approval {name}")
    return value


@dataclass(frozen=True, slots=True)
class ApprovalIntent:
    run_id: str
    root_run_id: str
    tool_call_id: str
    tool_name: str
    arguments: Mapping[str, Any]
    preset_fingerprint: str
    binding_id: str
    binding_revision: str
    scope_id: str
    scope_revision: str
    effect: str
    schema_version: int = 1

    def __post_init__(self):
        if type(self.schema_version) is not int or self.schema_version != 1:
            raise ValueError("invalid approval intent version")
        for name in _INTENT_NAMES:
            _text(getattr(self, name), name)
        if self.effect not in {"write", "destructive"}:
            raise ValueError("invalid approval effect")
        if not isinstance(self.arguments, Mapping):
            raise TypeError("approval arguments must be an object")
        object.__setattr__(self, "arguments", _ARGUMENTS.validate_value(self.arguments))
        json_identity_digest(self.to_mapping())

    def to_mapping(self):
        return {"schemaVersion": self.schema_version, "arguments": thaw_json_mapping(self.arguments),
                **{wire: getattr(self, name) for name, wire in _INTENT_NAMES.items()}}

    @property
    def digest(self):
        return json_identity_digest({"profile": APPROVAL_INTENT_PROFILE, "intent": self.to_mapping()})

    @classmethod
    def from_mapping(cls, value):
        if not isinstance(value, Mapping) or set(value) != {"schemaVersion", "arguments", *_INTENT_NAMES.values()}:
            raise ValueError("invalid approval intent fields")
        return cls(schema_version=value["schemaVersion"], arguments=value["arguments"],
                   **{name: value[wire] for name, wire in _INTENT_NAMES.items()})


_INTENT_NAMES = {"run_id": "runId", "root_run_id": "rootRunId", "tool_call_id": "toolCallId",
                 "tool_name": "toolName", "preset_fingerprint": "presetFingerprint", "binding_id": "bindingId",
                 "binding_revision": "bindingRevision", "scope_id": "scopeId", "scope_revision": "scopeRevision", "effect": "effect"}


@dataclass(frozen=True, slots=True)
class ApprovalDecisionCommand:
    approval_id: str
    expected_revision: int
    intent_digest: str
    command_key: str
    decision: str

    def __post_init__(self):
        for name in ("approval_id", "intent_digest", "command_key"):
            _text(getattr(self, name), name)
        _integer(self.expected_revision, "expected revision", 1)
        if self.decision not in {"approve", "reject"}:
            raise ValueError("invalid approval decision")

    def to_mapping(self):
        return {"approvalId": self.approval_id, "expectedRevision": self.expected_revision,
                "intentDigest": self.intent_digest, "commandKey": self.command_key, "decision": self.decision}


@dataclass(frozen=True, slots=True)
class ApprovalRecord:
    approval_id: str
    intent: ApprovalIntent
    revision: int
    status: str
    created_at_ms: int
    expires_at_ms: int
    decision_audit: Mapping[str, Any] = field(default_factory=dict, repr=False)

    def __post_init__(self):
        _text(self.approval_id, "id")
        if not isinstance(self.intent, ApprovalIntent):
            raise TypeError("approval intent is required")
        _integer(self.revision, "revision", 1)
        _integer(self.created_at_ms, "creation time")
        _integer(self.expires_at_ms, "expiry", self.created_at_ms + 1)
        if self.status not in APPROVAL_STATUSES:
            raise ValueError("invalid approval status")
        object.__setattr__(self, "decision_audit", freeze_json_mapping(self.decision_audit))
        audit = self.decision_audit
        if audit:
            if set(audit) != {"command", "principalId", "decidedAtMs", "revision"}:
                raise ValueError("invalid approval audit")
            command = audit["command"]
            if not isinstance(command, Mapping) or set(command) != {"approvalId", "expectedRevision", "intentDigest", "commandKey", "decision"}:
                raise ValueError("invalid approval audit command")
            ApprovalDecisionCommand(command["approvalId"], command["expectedRevision"], command["intentDigest"], command["commandKey"], command["decision"])
            _text(audit["principalId"], "principal")
            _integer(audit["decidedAtMs"], "decision time", self.created_at_ms)
            _integer(audit["revision"], "decision revision", 2)
            if (command["approvalId"] != self.approval_id or command["intentDigest"] != self.intent.digest
                    or command["expectedRevision"] + 1 != audit["revision"] or audit["revision"] > self.revision
                    or audit["decidedAtMs"] >= self.expires_at_ms or self.status == "pending"
                    or self.status in {"approved", "rejected"} and self.status != ("approved" if command["decision"] == "approve" else "rejected")):
                raise ValueError("conflicting approval audit")
        elif self.status in {"approved", "rejected"}:
            raise ValueError("approval decision audit is required")

    def to_mapping(self):
        return {"approvalId": self.approval_id, "intent": self.intent.to_mapping(), "intentDigest": self.intent.digest,
                "revision": self.revision, "status": self.status, "createdAtMs": self.created_at_ms,
                "expiresAtMs": self.expires_at_ms, "decisionAudit": thaw_json_mapping(self.decision_audit)}

    @classmethod
    def from_mapping(cls, value):
        if not isinstance(value, Mapping) or set(value) != {"approvalId", "intent", "intentDigest", "revision", "status", "createdAtMs", "expiresAtMs", "decisionAudit"}:
            raise ValueError("invalid approval record fields")
        intent = ApprovalIntent.from_mapping(value["intent"])
        if intent.digest != value["intentDigest"]:
            raise ValueError("conflicting approval intent digest")
        return cls(value["approvalId"], intent, value["revision"], value["status"], value["createdAtMs"], value["expiresAtMs"], value["decisionAudit"])


__all__ = ["APPROVAL_INTENT_PROFILE", "ApprovalIntent", "ApprovalDecisionCommand", "ApprovalRecord"]

File: typescript/src/approvals.ts
/** Immutable approval data; no record or digest is a dispatch permit. */
import { AgentError } from "./shared/errors.js";
import { StructuredOutputContract, jsonIdentityDigest } from "./structured.js";
import { copyJsonValue } from "./model/validation.js";
import type { JsonValue } from "./model/types.js";

export const APPROVAL_INTENT_PROFILE = "purra.approval-intent/v1";
export type ApprovalStatus = "pending" | "approved" | "rejected" | "expired" | "canceled";
export interface ApprovalIntentValue {
  readonly schemaVersion: 1;
  readonly runId: string;
  readonly rootRunId: string;
  readonly toolCallId: string;
  readonly toolName: string;
  readonly arguments: Readonly<Record<string, JsonValue>>;
  readonly presetFingerprint: string;
  readonly bindingId: string;
  readonly bindingRevision: string;
  readonly scopeId: string;
  readonly scopeRevision: string;
  readonly effect: "write" | "destructive";
}
const names = ["runId", "rootRunId", "toolCallId", "toolName", "presetFingerprint", "bindingId", "bindingRevision", "scopeId", "scopeRevision", "effect"] as const;
function text(value: unknown): asserts value is string {
  if (typeof value !== "string" || !value || value.trim() !== value || [...value].length > 1024) throw new TypeError("Invalid approval text");
}
function integer(value: unknown, minimum = 0): asserts value is number {
  if (!Number.isSafeInteger(value) || (value as number) < minimum) throw new TypeError("Invalid approval integer");
}
function keys(value: unknown, expected: readonly string[]): asserts value is Record<string, any> {
  if (value === null || typeof value !== "object" || Array.isArray(value)) throw new TypeError("Invalid approval fields");
  const actual = Object.keys(value).sort(), sorted = [...expected].sort();
  if (actual.length !== sorted.length || actual.some((key, index) => key !== sorted[index])) throw new TypeError("Invalid approval fields");
}

export class ApprovalIntent {
  private constructor(readonly value: ApprovalIntentValue, readonly digest: string) { Object.freeze(this); }
  static async create(input: ApprovalIntentValue): Promise<ApprovalIntent> {
    keys(input, ["schemaVersion", "arguments", ...names]);
    if (input.schemaVersion !== 1) throw new TypeError("Invalid approval intent version");
    for (const name of names) text(input[name]);
    if (!["write", "destructive"].includes(input.effect)) throw new TypeError("Invalid approval effect");
    // Snapshot before the first await; caller mutations cannot change the hashed intent.
    const copied = copyJsonValue(input as unknown as JsonValue) as unknown as ApprovalIntentValue;
    const contract = await StructuredOutputContract.create({ schemaId: "purra.approval-arguments", schemaVersion: "1", schema: { type: "object" } });
    contract.validateValue(copied.arguments);
    return new ApprovalIntent(copied, await jsonIdentityDigest({ profile: APPROVAL_INTENT_PROFILE, intent: copied }));
  }
}

export interface ApprovalDecisionCommand {
  readonly approvalId: string;
  readonly expectedRevision: number;
  readonly intentDigest: string;
  readonly commandKey: string;
  readonly decision: "approve" | "reject";
}
export function copyApprovalDecisionCommand(value: ApprovalDecisionCommand): ApprovalDecisionCommand {
  keys(value, ["approvalId", "expectedRevision", "intentDigest", "commandKey", "decision"]);
  text(value.approvalId); text(value.intentDigest); text(value.commandKey); integer(value.expectedRevision, 1);
  if (!["approve", "reject"].includes(value.decision)) throw new TypeError("Invalid approval decision");
  return Object.freeze({ ...value });
}
export interface ApprovalDecisionAudit {
  readonly command: ApprovalDecisionCommand;
  readonly principalId: string;
  readonly decidedAtMs: number;
  readonly revision: number;
}
export interface ApprovalRecord {
  readonly approvalId: string;
  readonly intent: ApprovalIntentValue;
  readonly intentDigest: string;
  readonly revision: number;
  readonly status: ApprovalStatus;
  readonly createdAtMs: number;
  readonly expiresAtMs: number;
  readonly decisionAudit: ApprovalDecisionAudit | Readonly<Record<string, never>>;
}
export async function copyApprovalRecord(input: ApprovalRecord): Promise<ApprovalRecord> {
  keys(input, ["approvalId", "intent", "intentDigest", "revision", "status", "createdAtMs", "expiresAtMs", "decisionAudit"]);
  const value = copyJsonValue(input as unknown as JsonValue) as unknown as ApprovalRecord;
  text(value.approvalId); integer(value.revision, 1); integer(value.createdAtMs); integer(value.expiresAtMs, value.createdAtMs + 1);
  if (!["pending", "approved", "rejected", "expired", "canceled"].includes(value.status)) throw new TypeError("Invalid approval status");
  const intent = await ApprovalIntent.create(value.intent);
  if (intent.digest !== value.intentDigest) throw new TypeError("Conflicting approval digest");
  const audit = value.decisionAudit;
  if (Object.keys(audit).length) {
    keys(audit, ["command", "principalId", "decidedAtMs", "revision"]);
    const command = copyApprovalDecisionCommand(audit.command);
    text(audit.principalId); integer(audit.decidedAtMs, value.createdAtMs); integer(audit.revision, 2);
    if (command.approvalId !== value.approvalId || command.intentDigest !== value.intentDigest
      || command.expectedRevision + 1 !== audit.revision || audit.revision > value.revision
      || audit.decidedAtMs >= value.expiresAtMs || value.status === "pending"
      || ["approved", "rejected"].includes(value.status) && value.status !== (command.decision === "approve" ? "approved" : "rejected")) throw new TypeError("Conflicting approval audit");
  } else {
    keys(audit, []);
    if (["approved", "rejected"].includes(value.status)) throw new TypeError("Approval decision audit is required");
  }
  return value;
}

/** Durable wait control flow, distinct from tool failure and user clarification. */
export class ApprovalRequired extends AgentError {
  constructor(readonly runId: string, readonly approvalId: string) {
    super("approval_required", "Run is waiting for approval");
    this.name = "ApprovalRequired";
  }
}
